File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/gql/staking.py
from human_protocol_sdk.filter import StakersFilter

STAKER_FRAGMENT = """
fragment StakerFields on Staker {
    id
    address
    stakedAmount
    lockedAmount
    withdrawnAmount
    slashedAmount
    lockedUntilTimestamp
    lastDepositTimestamp
}
"""


def get_stakers_query(filter: StakersFilter) -> str:
    where_fields = []
    if filter.min_staked_amount:
        where_fields.append("stakedAmount_gte: $minStakedAmount")
    if filter.max_staked_amount:
        where_fields.append("stakedAmount_lte: $maxStakedAmount")
    if filter.min_locked_amount:
        where_fields.append("lockedAmount_gte: $minLockedAmount")
    if filter.max_locked_amount:
        where_fields.append("lockedAmount_lte: $maxLockedAmount")
    if filter.min_withdrawn_amount:
        where_fields.append("withdrawnAmount_gte: $minWithdrawnAmount")
    if filter.max_withdrawn_amount:
        where_fields.append("withdrawnAmount_lte: $maxWithdrawnAmount")
    if filter.min_slashed_amount:
        where_fields.append("slashedAmount_gte: $minSlashedAmount")
    if filter.max_slashed_amount:
        where_fields.append("slashedAmount_lte: $maxSlashedAmount")

    where_clause = f"where: {{ {', '.join(where_fields)} }}" if where_fields else ""

    return f"""
query GetStakers(
    $minStakedAmount: BigInt
    $maxStakedAmount: BigInt
    $minLockedAmount: BigInt
    $maxLockedAmount: BigInt
    $minWithdrawnAmount: BigInt
    $maxWithdrawnAmount: BigInt
    $minSlashedAmount: BigInt
    $maxSlashedAmount: BigInt
    $orderBy: Staker_orderBy
    $orderDirection: OrderDirection
    $first: Int
    $skip: Int
) {{
    stakers(
        {where_clause}
        orderBy: $orderBy
        orderDirection: $orderDirection
        first: $first
        skip: $skip
    ) {{
        ...StakerFields
    }}
}}
{STAKER_FRAGMENT}
"""


def get_staker_query() -> str:
    return f"""
query GetStaker($id: String!) {{
    staker(id: $id) {{
        ...StakerFields
    }}
}}
{STAKER_FRAGMENT}
"""

File: packages/sdk/typescript/human-protocol-sdk/src/graphql/queries/staking.ts
import gql from 'graphql-tag';
import { IStakersFilter } from '../../interfaces';

const STAKER_FRAGMENT = gql`
  fragment StakerFields on Staker {
    id
    address
    stakedAmount
    lockedAmount
    withdrawnAmount
    slashedAmount
    lockedUntilTimestamp
    lastDepositTimestamp
  }
`;

export const GET_STAKERS_QUERY = (filter: IStakersFilter) => {
  const {
    minStakedAmount,
    maxStakedAmount,
    minLockedAmount,
    maxLockedAmount,
    minWithdrawnAmount,
    maxWithdrawnAmount,
    minSlashedAmount,
    maxSlashedAmount,
  } = filter;

  const whereFields = [
    minStakedAmount ? `stakedAmount_gte: $minStakedAmount` : '',
    maxStakedAmount ? `stakedAmount_lte: $maxStakedAmount` : '',
    minLockedAmount ? `lockedAmount_gte: $minLockedAmount` : '',
    maxLockedAmount ? `lockedAmount_lte: $maxLockedAmount` : '',
    minWithdrawnAmount ? `withdrawnAmount_gte: $minWithdrawnAmount` : '',
    maxWithdrawnAmount ? `withdrawnAmount_lte: $maxWithdrawnAmount` : '',
    minSlashedAmount ? `slashedAmount_gte: $minSlashedAmount` : '',
    maxSlashedAmount ? `slashedAmount_lte: $maxSlashedAmount` : '',
  ].filter(Boolean);

  const WHERE_CLAUSE = whereFields.length
    ? `where: { ${whereFields.join(', ')} }`
    : '';

  return gql`
    query getStakers(
      $minStakedAmount: BigInt
      $maxStakedAmount: BigInt
      $minLockedAmount: BigInt
      $maxLockedAmount: BigInt
      $minWithdrawnAmount: BigInt
      $maxWithdrawnAmount: BigInt
      $minSlashedAmount: BigInt
      $maxSlashedAmount: BigInt
      $orderBy: String
      $orderDirection: OrderDirection
      $first: Int
      $skip: Int
    ) {
      stakers(
        ${WHERE_CLAUSE}
        orderBy: $orderBy
        orderDirection: $orderDirection
        first: $first
        skip: $skip
      ) {
        ...StakerFields
      }
    }
    ${STAKER_FRAGMENT}
  `;
};

export const GET_STAKER_BY_ADDRESS_QUERY = gql`
  query getStaker($id: String!) {
    staker(id: $id) {
      ...StakerFields
    }
  }
  ${STAKER_FRAGMENT}
`;
